aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp

```diff
@@ -309,6 +309,29 @@ inline at::Tensor pack_weight_to_onednn_tensor(
   return packed_weight;
 }
 
+inline at::Tensor pack_weight_to_fp16_onednn_tensor(
+    at::Tensor& weight,
+    std::optional<torch::List<int64_t>>& input_shape) {
+  weight = at::_saturate_weight_to_fp16(weight);
+  std::vector<int64_t> w_dims = weight.sizes().vec();
+  ideep::tensor wei = ideep::tensor({w_dims, dnnl::memory::data_type::f32}, weight.data_ptr());
+  wei.transpose_(0, 1); // oneDNN requires transposed weight
+  ideep::dims input_dims = input_shape.has_value() ? input_shape.value().vec() : ideep::dims();
+  // Onednn does not support f32f16f32 matmul, so we need to convert weight to f32 before compute
+  // Therefore, we query weight layout with f32 dtype and store weight in f16 with the same layout
+  auto w_desc = ideep::matmul_forward::expected_weights_desc(
+      wei.get_dims(), input_dims, dnnl::memory::data_type::f32, dnnl::memory::data_type::f32);
+  w_desc = w_desc.to_type(ideep::data_type::f16);
+  ideep::tensor expected_weight(w_desc);
```

Comment:
Not sure if this would work well - you asked onednn for a memory descriptor of fp32 while then change it to fp16 to reorder to, seems not a common way of invoking the reorder API. Perhaps you can just use the plain layout since you are not using the blocked fp16 tensor anyway.

Reply:
Thanks. It works on latest Xeon but fails on older platforms. I have modified this part and now we return plain layout directly.